cmd/registry/cmd/upload/bulk/protos.go

```diff
@@ -289,10 +305,117 @@ func (task *uploadProtoTask) fileName() string {
 
 func (task *uploadProtoTask) zipContents() ([]byte, error) {
 	prefix := task.directory + "/"
-	contents, err := core.ZipArchiveOfPath(task.path, prefix, true)
+
+	// Get the proto files in the main directory.
+	protos, err := localProtos(task.path, prefix)
+	if err != nil {
+		return nil, err
+	}
+
+	// Compile the listed protos to get their dependencies.
+	count := len(protos)
+	for count > 0 {
+		protos, err = referencedProtos(protos, task.directory)
+		if err != nil {
+			return nil, err
+		}
+		if len(protos) == count {
+			break
+		}
+		count = len(protos)
+	}
```

Comment:
I'm having a hard time understanding this loop. First, I think the following is equivalent? 

```
for count := len(protos); count > 0; count = len(protos) {
	protos, err = referencedProtos(protos, task.directory)
	if err != nil {
		return nil, err
	}
	if len(protos) == count {
		break
	}
}
```

The exit conditions are still confusing me though. `count` is the previous length of `protos`? Get more referenced protos as long as the previous one wasn't empty, and stop if the current fetch has the same size as the previous? Why?

Reply:
That's all true -- I wrote this loop because the call to protoc in referencedProtos wasn't returning the dependencies of dependencies.... so I found that I had to iteratively run it to be sure I had everything.

Later I found that the `--include_imports` option seemed to make this work in one pass, but I left this loop here to verify that this was true (and not trusting my own interpretation). But the loop can probably be removed. (?)

The zero length check is present because I found a directory that appeared to be an API but that had no protos, and this caused protoc to fail. This change allows uploading to proceed (we upload a zip file of metadata only) and leaves the question of whether or not we *should* do that for later.

Reply:
I've replaced this with a non-iterated call to `referencedProtos` and verified it with the current `googleapis` protos.